File: packages/astro-identity-gen/src/polygon.ts
export type EdgeStyle = "flat" | "spikey" | "scalloped" | "inverse-scalloped";

export const edgeStyles: EdgeStyle[] = [
  "flat",
  "spikey",
  "scalloped",
  "inverse-scalloped",
];

export interface PolygonParams {
  /** Number of sides (3–8). */
  sides: number;
  /** Edge treatment style. */
  edgeStyle: EdgeStyle;
  /** Rotation in radians. */
  rotation: number;
  /** Radius as a fraction of half the viewbox (0.5–0.95). */
  radius: number;
  /** For spikey: inner radius ratio (0.3–0.7 of radius). */
  spikeDepth: number;
  /** For scalloped/inverse-scalloped: how far control points bow (0.2–0.5). */
  curveAmount: number;
}

interface Point {
  x: number;
  y: number;
}

/** Get the vertices of a regular polygon centered at (cx, cy). */
function regularVertices(
  sides: number,
  cx: number,
  cy: number,
  r: number,
  rotation: number,
): Point[] {
  const pts: Point[] = [];
  for (let i = 0; i < sides; i++) {
    const angle = rotation + (2 * Math.PI * i) / sides;
    pts.push({ x: cx + r * Math.cos(angle), y: cy + r * Math.sin(angle) });
  }
  return pts;
}

/** Build an SVG path string for a flat (straight-edged) polygon. */
function flatPath(vertices: Point[]): string {
  const [first, ...rest] = vertices;
  return (
    `M ${first.x} ${first.y} ` +
    rest.map((p) => `L ${p.x} ${p.y}`).join(" ") +
    " Z"
  );
}

/** Build an SVG path for a star / spikey polygon. */
function spikeyPath(
  sides: number,
  cx: number,
  cy: number,
  outerR: number,
  innerR: number,
  rotation: number,
): string {
  const pts: Point[] = [];
  for (let i = 0; i < sides; i++) {
    const outerAngle = rotation + (2 * Math.PI * i) / sides;
    pts.push({
      x: cx + outerR * Math.cos(outerAngle),
      y: cy + outerR * Math.sin(outerAngle),
    });
    const innerAngle = outerAngle + Math.PI / sides;
    pts.push({
      x: cx + innerR * Math.cos(innerAngle),
      y: cy + innerR * Math.sin(innerAngle),
    });
  }
  const [first, ...rest] = pts;
  return (
    `M ${first.x} ${first.y} ` +
    rest.map((p) => `L ${p.x} ${p.y}`).join(" ") +
    " Z"
  );
}

/** Build an SVG path with scalloped (outward-curved) or inverse-scalloped (inward-curved) edges. */
function scallopedPath(
  vertices: Point[],
  cx: number,
  cy: number,
  curveAmount: number,
  inward: boolean,
): string {
  const n = vertices.length;
  const parts: string[] = [`M ${vertices[0].x} ${vertices[0].y}`];

  for (let i = 0; i < n; i++) {
    const a = vertices[i];
    const b = vertices[(i + 1) % n];
    // Midpoint of the edge
    const mx = (a.x + b.x) / 2;
    const my = (a.y + b.y) / 2;
    // Direction from center to midpoint
    const dx = mx - cx;
    const dy = my - cy;
    const dist = Math.sqrt(dx * dx + dy * dy);
    const nx = dx / dist;
    const ny = dy / dist;
    // Push control point outward or inward
    const sign = inward ? -1 : 1;
    const offset = curveAmount * dist;
    const cpx = mx + sign * nx * offset;
    const cpy = my + sign * ny * offset;
    parts.push(`Q ${cpx} ${cpy} ${b.x} ${b.y}`);
  }

  parts.push("Z");
  return parts.join(" ");
}

/** Generate a full SVG path `d` attribute for the given polygon parameters. */
export function buildPolygonPath(params: PolygonParams, size: number): string {
  const cx = size / 2;
  const cy = size / 2;
  const r = params.radius * (size / 2);

  switch (params.edgeStyle) {
    case "flat": {
      const verts = regularVertices(params.sides, cx, cy, r, params.rotation);
      return flatPath(verts);
    }
    case "spikey": {
      const innerR = r * params.spikeDepth;
      return spikeyPath(params.sides, cx, cy, r, innerR, params.rotation);
    }
    case "scalloped": {
      const verts = regularVertices(params.sides, cx, cy, r, params.rotation);
      return scallopedPath(verts, cx, cy, params.curveAmount, false);
    }
    case "inverse-scalloped": {
      const verts = regularVertices(params.sides, cx, cy, r, params.rotation);
      return scallopedPath(verts, cx, cy, params.curveAmount, true);
    }
  }
}

File: apps/astro-server/internal/identitygen/polygon.go
package identitygen

import (
	"math"
	"strconv"
	"strings"
)

// EdgeStyle names a polygon edge treatment.
type EdgeStyle string

const (
	EdgeFlat             EdgeStyle = "flat"
	EdgeSpikey           EdgeStyle = "spikey"
	EdgeScalloped        EdgeStyle = "scalloped"
	EdgeInverseScalloped EdgeStyle = "inverse-scalloped"
)

// edgeStyles is the ordered list generateIdentity picks from. Order is
// load-bearing — rng indexes into it.
var edgeStyles = []EdgeStyle{
	EdgeFlat,
	EdgeSpikey,
	EdgeScalloped,
	EdgeInverseScalloped,
}

// PolygonParams describes the polygon layer of an identity.
type PolygonParams struct {
	Sides       int       // 3–8
	EdgeStyle   EdgeStyle
	Rotation    float64   // radians
	Radius      float64   // fraction of half-viewBox
	SpikeDepth  float64   // inner-radius ratio for spikey
	CurveAmount float64   // control-point offset for scalloped
}

type point struct {
	x, y float64
}

// regularVertices returns the vertices of a regular polygon centered at (cx, cy).
func regularVertices(sides int, cx, cy, r, rotation float64) []point {
	pts := make([]point, sides)
	for i := range sides {
		angle := rotation + 2*math.Pi*float64(i)/float64(sides)
		pts[i] = point{x: cx + r*math.Cos(angle), y: cy + r*math.Sin(angle)}
	}
	return pts
}

func flatPath(vertices []point) string {
	var b strings.Builder
	b.WriteString("M ")
	writeFloat(&b, vertices[0].x)
	b.WriteByte(' ')
	writeFloat(&b, vertices[0].y)
	for _, p := range vertices[1:] {
		b.WriteString(" L ")
		writeFloat(&b, p.x)
		b.WriteByte(' ')
		writeFloat(&b, p.y)
	}
	b.WriteString(" Z")
	return b.String()
}

// spikeyPath alternates outer and inner radii to build a star-like shape.
func spikeyPath(sides int, cx, cy, outerR, innerR, rotation float64) string {
	pts := make([]point, 0, sides*2)
	for i := range sides {
		outerAngle := rotation + 2*math.Pi*float64(i)/float64(sides)
		pts = append(pts, point{
			x: cx + outerR*math.Cos(outerAngle),
			y: cy + outerR*math.Sin(outerAngle),
		})
		innerAngle := outerAngle + math.Pi/float64(sides)
		pts = append(pts, point{
			x: cx + innerR*math.Cos(innerAngle),
			y: cy + innerR*math.Sin(innerAngle),
		})
	}
	return flatPath(pts)
}

// scallopedPath draws Bezier arcs with control points pushed outward (inward=false)
// or inward (inward=true) relative to the polygon center.
func scallopedPath(vertices []point, cx, cy, curveAmount float64, inward bool) string {
	var b strings.Builder
	b.WriteString("M ")
	writeFloat(&b, vertices[0].x)
	b.WriteByte(' ')
	writeFloat(&b, vertices[0].y)

	n := len(vertices)
	sign := 1.0
	if inward {
		sign = -1
	}
	for i := range n {
		a := vertices[i]
		bp := vertices[(i+1)%n]
		mx := (a.x + bp.x) / 2
		my := (a.y + bp.y) / 2
		dx := mx - cx
		dy := my - cy
		dist := math.Sqrt(dx*dx + dy*dy)
		nx := dx / dist
		ny := dy / dist
		offset := curveAmount * dist
		cpx := mx + sign*nx*offset
		cpy := my + sign*ny*offset
		b.WriteString(" Q ")
		writeFloat(&b, cpx)
		b.WriteByte(' ')
		writeFloat(&b, cpy)
		b.WriteByte(' ')
		writeFloat(&b, bp.x)
		b.WriteByte(' ')
		writeFloat(&b, bp.y)
	}

	b.WriteString(" Z")
	return b.String()
}

// buildPolygonPath returns the `d` attribute for a polygon defined by params
// drawn inside a size×size viewBox.
func buildPolygonPath(params PolygonParams, size int) string {
	sz := float64(size)
	cx := sz / 2
	cy := sz / 2
	r := params.Radius * sz / 2

	switch params.EdgeStyle {
	case EdgeFlat:
		return flatPath(regularVertices(params.Sides, cx, cy, r, params.Rotation))
	case EdgeSpikey:
		return spikeyPath(params.Sides, cx, cy, r, r*params.SpikeDepth, params.Rotation)
	case EdgeScalloped:
		return scallopedPath(regularVertices(params.Sides, cx, cy, r, params.Rotation), cx, cy, params.CurveAmount, false)
	case EdgeInverseScalloped:
		return scallopedPath(regularVertices(params.Sides, cx, cy, r, params.Rotation), cx, cy, params.CurveAmount, true)
	}
	return ""
}

// writeFloat formats a float coordinate into b using Go's shortest round-trip
// decimal representation. Formatting need not match the TS implementation —
// SVG renderers accept any valid decimal.
func writeFloat(b *strings.Builder, v float64) {
	b.WriteString(strconv.FormatFloat(v, 'f', -1, 64))
}
